### PR TITLE
[SYCL] Add identityless reduction test cases

### PR DESCRIPTION
**Pull request by steffenlarsen**
This commit adds test cases to existing tests for testing reductions where the identity is unspecified and unknown.

### CODE REVIEW

File: SYCL/Reduction/reduction_range_1d_rw.cpp [hunk 1]
```diff
@@ -12,6 +12,16 @@ using namespace sycl;
 
 int NumErrors = 0;
 
+template <typename Name, typename T, class BinaryOperation>
+void tests(queue &Q, T Identity, T Init, BinaryOperation BOp, range<1> Range) {
+  NumErrors += test<Name>(Q, Identity, Init, BOp, Range);
+}
```
Comment (aelovikov-intel) [inline SYCL/Reduction/reduction_range_1d_rw.cpp]: nit, if possible:
```
template <typename Name, typename... ArgTys>
void tests(ArgTys &&Args) {
  NumErrors += test<Name>(std::forward<ArgTys>(Args)...);
}
```

Same in other places.

File: SYCL/Reduction/reduction_range_1d_rw.cpp [hunk 2]
```diff
@@ -20,29 +30,40 @@ int main() {
 
   constexpr access::mode RW = access::mode::read_write;
   // Fast-reduce and Fast-atomics. Try various range types/sizes.
-  test<class A1, int>(Q, 0, 99, std::plus<int>{}, range<1>(1));
-  test<class A2, int>(Q, 0, 99, std::plus<>{}, range<1>(2));
-  test<class A3, int>(Q, 0, 99, std::plus<>{}, range<1>(7));
-  test<class A4, int>(Q, 0, 99, std::plus<>{}, range<1>(64));
-  test<class A5, int>(Q, 0, 99, std::plus<>{}, range<1>(MaxWGSize * 2));
-  test<class A6, int>(Q, 0, 99, std::plus<>{}, range<1>(MaxWGSize * 2 + 5));
+  tests<class A1, int>(Q, 0, 99, std::plus<int>{}, range<1>(1));
```
Comment (aelovikov-intel) [inline SYCL/Reduction/reduction_range_1d_rw.cpp:28]: How did it work with `test` ? Were errors ignored?

File: SYCL/Reduction/reduction_range_2d_rw.cpp [hunk 3]
```diff
@@ -48,6 +53,22 @@ int main() {
 
   tests<class C1>(Q, CustomVec<long long>(0), CustomVec<long long>(99),
                   CustomVecPlus<long long>{}, range<2>{33, MaxWGSize});
+  tests<class C2>(Q, CustomVec<long long>(99), CustomVecPlus<long long>{},
```
Comment (aelovikov-intel) [inline SYCL/Reduction/reduction_range_2d_rw.cpp]: Do we really need to keep the one *with* the identity param? Especially given the new test cases below to test that path.

File: SYCL/Reduction/reduction_range_1d_rw.cpp (hunk 2 above)
Comment (steffenlarsen) [reply SYCL/Reduction/reduction_range_1d_rw.cpp:28]: It looks like the results were indeed being ignored. They shouldn't be anymore though.

File: SYCL/Reduction/reduction_range_2d_rw.cpp (hunk 3 above)
Comment (steffenlarsen) [reply SYCL/Reduction/reduction_range_2d_rw.cpp]: I think it makes sense to test for custom types to make sure that runtime identities also work correctly for custom types.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by steffenlarsen** [comment]
Some test cases are disabled temporarily due to a known bug in reductions. A patch for this will follow after https://github.com/intel/llvm/pull/8562.

**Comment by steffenlarsen** [comment]
Verified with https://github.com/intel/llvm/pull/8562.

**Pull request merged**